### ISSUE TITLE
The heuristic encoding approach section

### ISSUE THREAD

**Issue by allegroCoder** [tuura/scenco-article#10]
Below the things that need to be done:
- [ ] Writing a short introduction about the aim of the section -- how it is divided.
- [ ] Writing the last subsection `Support for encoding constraints`.
- [ ] Proof reading the existing sections
- [ ] Checking for any inconsistencies with the added sections -- benchmarks -- terms
- [ ] Double checking Table I

**Comment by allegroCoder** [comment]
I have made a script for running the heuristic encoding over all the benchmarks present in `graph-families` repo, and adjusted `scenco` to output only the useful information (area - gate count) for the script purposes.

Now, we should be able to compare quickly different cost functions, if we found any. :)

@snowleopard What about setting up a meeting for discussing possible cost functions that we can try?

**Comment by snowleopard** [comment]
@allegroCoder How long does it take to run the whole benchmark suite?

I think we need to automate running the benchmarks by putting your script into CI, say into `graph-families` repository, and whenever we change the benchmarks or `scenco` sources, we need to regenerate the table with comparisons and release it (just like we do with PDF generation for papers).

Can you do that? Then we are always sure that the benchmarks results are accurate and correspond exactly to our benchmarks and code. After that it will be easy to experiment with cost functions.

**Comment by allegroCoder** [comment]
@snowleopard It takes about 13 minutes in my laptop.

So, you are asking to automatically update the table in `graph-families`, whenever either the code in `scenco-workcraft` code is modified or/and when any other benchmark is added into the repo?

To do that, I might need to insert inside `graph-families` a folder with the `scenco-workcraft` code (is there any other way to capture the modification of the code of another repo?), accordingly modified to output just what we need and the lib we want to try (`workcraft-no-async.lib` so far, might be even a set of libs). Then, the table, which will contain the results for all of the benchmarks (`md` format?), will be deployed as we did for the pdf generation. This will enclose the result of the last run of the heuristic algorithm which we can then compare. Would you be happy with this solution?

I am not sure about this flow though. Would it not be better to manually update the table when we find an interesting cost function (better or worse)? Moreover, the results of the table will be changing at every PR, even though the cost function remains the same (becoming inconsistent for the table present in `scenco-article` which might ne to be updated at every iteration as well). As we discussed at the Buttery, at some point it would be nice to have an automated way to verify that `Scenco` works fine in `Workcraft`, but I am not sure if I would automate the table generation. Let me know your thoughts :)

**Comment by snowleopard** [comment]
As discussed offline, I think it would be useful and not too difficult to automate this, so we always have up-to-date benchmark table, matching both the data (in `graph-families`) and code (in `scenco`). 

I think we only need to update this on push to the master branch.

**Comment by allegroCoder** [comment]
Only one note : the SAT-Based encoding is not integrated into `Scenco`. Therefore, I will update the table at every push to the master branch for the `Sequential`, `Random`, `Heuristic` and `Single-literal` encodings. At some point, it would be nice to have all of the encoding methods automatically updated into a table, in that case we should probably rely on `Workcraft` though.

At least we will be able to compare the cost function at every push with this intermediate solution. What do you think @snowleopard ?

**Comment by snowleopard** [comment]
Good point! The SAT-based method is my problem, I'll need to take care of this and provide a standalone implementation at some me point. Don't worry about it now.

**Comment by snowleopard** [comment]
Regarding comparing different cost functions, you don't necessarily need to push: you can just run the script locally to generate the tables.

**Comment by allegroCoder** [comment]
I was thinking actually of automating the generation of the results into the table in Travis. In the case that the script is run locally, I think it would be even easier to automate. I am just not sure if it is possible to deploy a document into a repo, from a different repo (in such a way to have both `graph-families` and `scenco-workcraft` with the same up-to-date table). I believe so, I will give it a try!

**Comment by allegroCoder** [comment]
I am done with the automation of the table generation. I will now automate the deployment in travis process ASAP  :+1:
